src/builder-manifest.c

```diff
@@ -3777,37 +3843,44 @@ builder_manifest_install_dep (BuilderManifest *self,
                                    builder_context_get_arch (context));
 
   commit = flatpak_info (opt_user, opt_installation, "--show-commit", ref, NULL);
-  installed = (commit != NULL);
 
-  args = g_ptr_array_new_with_free_func (g_free);
-  g_ptr_array_add (args, g_strdup ("flatpak"));
-  add_installation_args (args, opt_user, opt_installation);
-  if (installed)
+  if (commit != NULL)
     {
-      g_ptr_array_add (args, g_strdup ("update"));
-      g_ptr_array_add (args, g_strdup ("--subpath="));
+      g_print("Updating %s\n", ref);
+      if (builder_manifest_update_single_dep(ref, opt_user, opt_installation,
+                                             error))
+      {
+        return TRUE;
+      }
     }
   else
     {
-      g_ptr_array_add (args, g_strdup ("install"));
-      g_ptr_array_add (args, g_strdup (remote));
-    }
-
-  g_ptr_array_add (args, g_strdup ("-y"));
-  if (flatpak_version_check (1, 2, 0))
-    g_ptr_array_add (args, g_strdup ("--noninteractive"));
-
-  g_ptr_array_add (args, g_strdup (ref));
-  g_ptr_array_add (args, NULL);
-
-  if (!builder_maybe_host_spawnv (NULL, NULL, 0, error, (const char * const *)args->pdata))
-    {
-      g_autofree char *commandline = flatpak_quote_argv ((const char **)args->pdata);
-      g_prefix_error (error, "running `%s`: ", commandline);
-      return FALSE;
+      gboolean multiple_remotes = (*(remotes+1) != NULL);
+      for (const char *remote = *remotes; remote != NULL; remote = *(++remotes))
+	{
+	  g_autoptr(GError) current_error = NULL;
+	  if (multiple_remotes)
+	    {
+            g_print("Trying to install %s from %s\n", ref, remote);
+	    }
+	  else {
+	    g_print("Installing %s from %s\n", ref, remote);
+	  }
+	  if (builder_manifest_install_single_dep (ref, remote, opt_user, opt_installation,
+                                                   &current_error))
+	  {
+	    current_error = g_steal_pointer(error);
+	    return TRUE;
+	  }
+	  else if (*error == NULL)
+	    {
+	      *error = g_steal_pointer(&current_error);
+	    }
+	  if ((*error)->domain != G_SPAWN_EXIT_ERROR)
+	    return FALSE;
```

Comment:
The error handing here is a bit weird, as you're basically using the error remote arg as a local store for the error on the first error and then unsetting it when something later succeeds. I would recommend instead that you store the errors only in local variables and then call g_propagate_error (error, xxx_error) the line before you return FALSE.

Like, its not going to happen in this code, but in general you should always treat GError **error out variable as potentially NULL if the caller was not interested in error details.

Reply:
I agree the code is a bit weird and will change as you asked but if I've understood correctly, that's essentially the end result. Stealing nulls the original.

Reply:
Yeah, the result will be the same (if error is not NULL), but its good to have the code be more obvious.